modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/transaction/TransactionService.scala

```diff
@@ -0,0 +1,113 @@
+package org.tessellation.dag.l1.infrastructure.transaction
+
+import cats.data.{NonEmptyList, NonEmptySet}
+import cats.effect.{Ref, Sync}
+import cats.syntax.alternative._
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+import cats.syntax.show._
+
+import scala.collection.immutable.SortedSet
+
+import org.tessellation.dag.l1.domain.transaction.TransactionService
+import org.tessellation.dag.l1.infrastructure.{Success, UnexpectedStateShift, UpdateResult}
+import org.tessellation.dag.transaction.filter.Consecutive
+import org.tessellation.ext.collection.MapRefUtils._
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.transaction.{Transaction, TransactionOrdinal, TransactionReference}
+import org.tessellation.security.Hashed
+import org.tessellation.security.signature.Signed
+
+import io.chrisdavenport.mapref.MapRef
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object TransactionService {
+
+  def make[F[_]: Sync]: F[TransactionService[F]] =
+    for {
+      lastAccepted <- MapRef.ofConcurrentHashMap[F, Address, TransactionReference]()
+      waitingTransactions <- Ref.of[F, Map[Address, NonEmptySet[Signed[Transaction]]]](Map.empty)
+      transactionStorage = make(lastAccepted, waitingTransactions)
+    } yield transactionStorage
+
+  def make[F[_]: Sync](
+    lastAccepted: MapRef[F, Address, Option[TransactionReference]],
+    waitingTransactions: Ref[F, Map[Address, NonEmptySet[Signed[Transaction]]]]
+  ): TransactionService[F] =
+    new TransactionService[F] {
+      implicit val logger = Slf4jLogger.getLogger[F]
+
+      def isParentAccepted(transaction: Transaction): F[Boolean] =
+        (transaction.parent != TransactionReference.empty)
+          .guard[Option]
+          .fold(true.pure[F]) { _ =>
+            lastAccepted(transaction.source).get.map(_.contains(transaction.parent))
+          }
+
+      def getLastAcceptedTransactionRef(source: Address): F[TransactionReference] =
+        lastAccepted(source).get.map(_.getOrElse(TransactionReference.empty))
+
+      def accept(hashedTx: Hashed[Transaction]): F[Unit] = {
+        val source = hashedTx.signed.value.source
+        val reference = TransactionReference(hashedTx.hash, hashedTx.signed.value.ordinal)
+
+        lastAccepted(source)
+          .modify[UpdateResult] { lastAccepted =>
+            if (lastAccepted.isEmpty && reference.ordinal == TransactionOrdinal.first ||
+                lastAccepted.exists(_.ordinal.next == reference.ordinal))
+              (reference.some, Success)
+            else
+              (
+                lastAccepted,
+                UnexpectedStateShift(
+                  s"Transaction acceptance failed! Transaction not accepted in the correct order. source=${source.show} current=${lastAccepted.show} attempted=${reference.show}"
+                )
```

Comment:
IMO it should be a separate error. It's not a state shift but a domain error (missing parents, parents not accepted, or incorrect reference, etc)